lib/pdf_parser.rb

```diff
@@ -0,0 +1,159 @@
+    result = []
+    # use inject to iterate by adjacent pairs
+    blocks.inject do |top_block, bottom_block|
+      result << top_block
+      arabic_start = block_bottom_bound(top_block, page_map)
+      arabic_end = block_top_bound(bottom_block, page_map)
+      unless arabic_start.nil? || arabic_end.nil?
+        arabic = ocr_arabic_block(arabic_start[:word][:y_end],
+                                  arabic_end[:word][:y_start],
+                                  img_path)
+        result << arabic
+      end
+      # accumulate bottom block
+      bottom_block
+    end
+    result << blocks.last
+    result.join(' ')
+  end
+
+  # split the page text around non-latin blocks
+  def block_text_by_non_latin(page_text)
+    chars = page_text.split('')
+    # aggregate all of the latin text
+    result = chars.reduce([[]]) do |total, c|
+      if c.is_non_latin?
+        total.push [] unless total.last.empty?
+      elsif total.last.empty? && !c.is_whitespace?
+        # just out of non-latin
+        total.last.push c
+      elsif !total.last.empty?
+        # in latin
+        total.last.push c
+      end
+      total
+    end
+    result.map! &:join
+    result.reject! { |block| !block.strip.is_word? || block.nil? } # filtering
+    return result
+
+  end
+
+  # get the last word of the block
+  def block_bottom_bound(block, page_map)
+    block = block.split(' ')
+    matches = page_map.words.each_with_index.select do |word, _idx|
+      word[:word].downcase.include?(find_first_word(block.reverse).downcase)
+    end.map &:last
+
+    # contextually locate the top word
+    matches.map! do |match|
+      ocr_string = []
+      for i in (match / 2..match)
+        ocr_string << page_map.words[i][:word]
+      end
+      ocr_string = ocr_string.join(' ')
+      parsed_string = block[0..block.length].join(' ')
+      similarity = ocr_string.similar(parsed_string)
+      match = { similarity: similarity, word: page_map.words[match] }
+    end.sort_by! { |e| e[:similarity] }
+
+    result = matches.last # the best result
+    result[:block] = block.join(' ') unless result.nil?
+    result
+  end
+
+  # get the first word of a block
+  def block_top_bound(block, page_map)
+    block = block.split(' ')
+    # get the possible word locations
+    matches = page_map.words.each_with_index.select do |word, _idx|
+      word[:word].downcase.include?(find_first_word(block).downcase)
+    end.map &:last
+
+    # contextually locate the top word
+    matches.map! do |match|
+      ocr_string = []
+      page_map.words[match..match * 2].each { |word| ocr_string << word[:word] }
+      ocr_string = ocr_string.join(' ')
+      parsed_string = block[0..block.length].join(' ')
+      similarity = ocr_string.similar(parsed_string)
+      match = { similarity: similarity, word: page_map.words[match] }
+    end.sort_by! { |e| e[:similarity] }
+
+    result = matches.last # the best result
+    result[:block] = block.join(' ') unless result.nil?
+    result
```

Comment:
@mzagaja I was thinking along the same lines until I ran into a bug caused by what @ryanttb just described. I'll add an explicit return to make it more clear.